jax/core.py

```diff
@@ -639,21 +639,38 @@ def __repr__(self): return '*'
 identity_p.def_impl(lambda x: x)
 identity_p.def_custom_bind(lambda x: x)
 
-def concretization_err_msg(fun, context=None):
+class ConcretizationTypeError(TypeError): pass
+
+def raise_concretization_error(val, context=""):
+  msg = (f"Abstract tracer value encountered where a concrete value is expected ({context}).\n"
+          "Use `functools.partial` or transformation parameters such as `static_argnums` for `jit` "
```

Comment:
Using `functools.partial` and lexical closure are less good for `jit` because they can lead to more recompilation than `static_argnums` (by defining new function objects every time). How about we just recommend `static_argnums` here?